### PR TITLE
[NT-865] (2/2) Sign in with apple

### PR DESCRIPTION
**Pull request by Scollaco**
# 📲 What

- This is the second part of the SignIn with Apple work. It fetches the user from v1 once the sign in is successful and login the user into the Environment. 
- Handles errors if the `Sign In With Apple` flow fails.

# 🤔 Why

- This is a feature required by Apple and has to be implemented by June 2020.
[JIRA ticket](https://kickstarter.atlassian.net/browse/NT-865)

# 🛠 How

- After `appleSignInEvent` is completed, we extract `apiAccessToken` returned by the mutation.
- After `appleSignInEvent` is completed, we extract `userId` to fetch the user using a v1 API call.
- Once we have the new `User` and `apiAccessToken`, we create a AccessTokenEnvelope object and use it to login to the Environment.
- If `appleAuthorizationError`,` fetchUserEventError`, or `appleSignInEventError` emit, we show an alert with the error message.
- AuthenticationServices returns an error if the user taps the `cancel` button of the Sign In With Apple prompt. A filter was created to ignore this kind of errors to provide a better experience.
- Added tests

# 👀 See

signin-apple-flow

# ✅ Acceptance criteria

- [ ] Logout from the app and try to login/signup with Apple pointing to native's Hivequeen. If successful, the app should login normally.
- [ ] If an error occurred, an Alert should be presented with the error message
- [ ] Tap the `Continue with Apple` button and dismiss the modal by tapping `Cancel`. A error  message should **not** appear.

### CODE REVIEW

**Comment by nativeksr** [comment]
### SwiftFormat found issues:

| File | Rules |
| ---- | ----- |
| Library/ViewModels/LoginToutViewModel.swift | hoistPatternLet, indent |


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Library/ViewModels/LoginToutViewModelTests.swift [hunk 1]
```diff
@@ -588,6 +591,93 @@ final class LoginToutViewModelTests: TestCase {
...
+
+      self.logIntoEnvironment.assertDidNotEmitValue()
+
+      self.vm.inputs.appleAuthorizationDidSucceed(with: data)
+
+      self.scheduler.run()
+
+      let value = self.logIntoEnvironment.values.first
+
+      XCTAssertEqual(user, value?.user)
```
Comment (ifbarrera) [inline Library/ViewModels/LoginToutViewModelTests.swift:676]: Can we assert that `logIntoEnvironment` only fires once?

File: Library/ViewModels/LoginToutViewModel.swift [hunk 2]
```diff
@@ -203,10 +211,41 @@ public final class LoginToutViewModel: LoginToutViewModelType, LoginToutViewMode
...
+
+    let fetchUserEvent = userId
+      .switchMap { id in
+        AppEnvironment.current.apiService.fetchUser(userId: id)
+          .ksr_delay(AppEnvironment.current.apiDelayInterval, on: AppEnvironment.current.scheduler)
+          .prefix(SignalProducer([AppEnvironment.current.currentUser].compact()))
+          .materialize()
+      }
+
+    let logIntoEnvironmentWithApple = Signal.combineLatest(fetchUserEvent.values(), apiAccessToken)
```
[the remark this answers is not in the document]
Comment (Scollaco) [reply Library/ViewModels/LoginToutViewModel.swift:229]: Nice!

File: Library/ViewModels/LoginToutViewModel.swift [hunk 3]
```diff
@@ -203,10 +211,41 @@ public final class LoginToutViewModel: LoginToutViewModelType, LoginToutViewMode
...
+      .skipNil()
+
+    let apiAccessToken = appleSignInEvent.values()
+      .map(\.signInWithApple.apiAccessToken)
+
+    let fetchUserEvent = userId
+      .switchMap { id in
+        AppEnvironment.current.apiService.fetchUser(userId: id)
+          .ksr_delay(AppEnvironment.current.apiDelayInterval, on: AppEnvironment.current.scheduler)
+          .prefix(SignalProducer([AppEnvironment.current.currentUser].compact()))
```
[the remark this answers is not in the document]
Comment (Scollaco) [reply Library/ViewModels/LoginToutViewModel.swift:225]: Copy pasta 😛

**Pull request merged**